### PR TITLE
stream: provide mapMaterializedValue for Graph

### PR DESCRIPTION
**Pull request by eyalfa**
implement #28501 

1. added a _static_ `mapMaterializedValue` to class `Graph`
2. added the `mapMaterializedValue` method to class `Graph` via an extension method.

see discussion in the issue for explanations why this method could be added directly to class `Graph` (spoiler, java inter-op)

### CODE REVIEW

**Comment by eyalfa** [comment]
cc: @johanandren , @jrudolph

**Comment by johanandren** [comment]
OK TO TEST

File: akka-stream/src/main/scala/akka/stream/Graph.scala [hunk 1]
```diff
@@ -69,6 +70,15 @@ trait Graph[+S <: Shape, +M] {
   def addAttributes(attr: Attributes): Graph[S, M] = withAttributes(traversalBuilder.attributes and attr)
 }
 
+object Graph {
+  def mapMaterializedValue[S <: Shape, M1, M2](g: Graph[S, M1])(f: M1 => M2): Graph[S, M2] = {
+    new GenericGraph(g.shape, g.traversalBuilder).mapMaterializedValue(f)
+  }
+  implicit class GraphMapMatVal[S <: Shape, M](self: Graph[S, M]) {
```
Comment (johanandren) [inline akka-stream/src/main/scala/akka/stream/Graph.scala]: `final`

File: akka-stream/src/main/scala/akka/stream/Graph.scala [hunk 2]
```diff
@@ -69,6 +70,15 @@ trait Graph[+S <: Shape, +M] {
   def addAttributes(attr: Attributes): Graph[S, M] = withAttributes(traversalBuilder.attributes and attr)
 }
 
+object Graph {
+  def mapMaterializedValue[S <: Shape, M1, M2](g: Graph[S, M1])(f: M1 => M2): Graph[S, M2] = {
```
Comment (johanandren) [inline akka-stream/src/main/scala/akka/stream/Graph.scala]: Scaladoc like on the other `mapMaterializedValue`s on both of these, perhaps also mention this is Java API and the implicit one Scala API?

**Review by johanandren — COMMENTED** [review]
Looking good, some docs and  think this is ready to be merged.

**Review by johanandren — APPROVED** [review]
LGTM

**Comment by eyalfa** [comment]
@johanandren , can u please ask another team member to review?

File: akka-stream-tests/src/test/java/akka/stream/javadsl/GraphDslTest.java [hunk 3]
```diff
@@ -216,4 +217,14 @@ public void beAbleToConstructClosedGraphFromList() throws Exception {
     assertEquals("bx", result.get(1).toCompletableFuture().get(1, TimeUnit.SECONDS));
     assertEquals("cx", result.get(2).toCompletableFuture().get(1, TimeUnit.SECONDS));
   }
+
+  @Test
+  public void canUseMapMaterializedValueOnGraphs() {
+    Graph<SourceShape<Object>, NotUsed> srcGraph = Source.empty();
+    Graph<SourceShape<Object>, Pair> mappedMatValueSrcGraph =
+        Graph.mapMaterializedValue(srcGraph, notUsed -> new Pair(notUsed, notUsed));
```
Comment (jrudolph) [inline akka-stream-tests/src/test/java/akka/stream/javadsl/GraphDslTest.java]: A bit weird type to map to but I guess it doesn't matter :)

File: akka-stream/src/main/scala/akka/stream/Graph.scala [hunk 4]
```diff
@@ -69,6 +70,35 @@ trait Graph[+S <: Shape, +M] {
...
+   * @param f function to map the graph's materialized value
+   * @return a graph with same semantics as the given graph, except from the materialized value which is mapped using f.
+   */
+  def mapMaterializedValue[S <: Shape, M1, M2](g: Graph[S, M1])(f: M1 => M2): Graph[S, M2] = {
+    new GenericGraph(g.shape, g.traversalBuilder).mapMaterializedValue(f)
+  }
+
+  /**
+   * Scala API, see https://github.com/akka/akka/issues/28501 for discussion why this can't be an instance method on class Graph.
+   * @param self the graph whose material value will be mapped
```
Comment (jrudolph) [inline akka-stream/src/main/scala/akka/stream/Graph.scala]:
```suggestion
   * @param self the graph whose materialized value will be mapped
```

Comment (jrudolph) [reply akka-stream/src/main/scala/akka/stream/Graph.scala]: Not sure we need scaladoc here, seems unlikely you would end up here? But doesn't hurt as well.

File: akka-stream/src/main/scala/akka/stream/Graph.scala [hunk 5]
```diff
@@ -69,6 +70,35 @@ trait Graph[+S <: Shape, +M] {
...
+   */
+  final implicit class GraphMapMatVal[S <: Shape, M](self: Graph[S, M]) {
+
+    /**
+     * Transform the materialized value of this Flow, leaving all other properties as they were.
```
Comment (jrudolph) [inline akka-stream/src/main/scala/akka/stream/Graph.scala]:
```suggestion
     * Transform the materialized value of this Graph, leaving all other properties as they were.
```

File: akka-stream/src/main/scala/akka/stream/Graph.scala (hunk 2 above)
Comment (jrudolph) [reply akka-stream/src/main/scala/akka/stream/Graph.scala]: Minor: no braces needed here.

File: akka-stream/src/main/scala/akka/stream/scaladsl/Graph.scala [hunk 6]
```diff
@@ -37,6 +37,10 @@ private[stream] final class GenericGraph[S <: Shape, Mat](
 
   override def withAttributes(attr: Attributes): Graph[S, Mat] =
     new GenericGraphWithChangedAttributes(shape, traversalBuilder, attr)
+
+  def mapMaterializedValue[Mat2](f: Mat => Mat2): GenericGraph[S, Mat2] = {
```
Comment (jrudolph) [inline akka-stream/src/main/scala/akka/stream/scaladsl/Graph.scala]: Minor: no braces needed here.

**Review by jrudolph — APPROVED** [review]
LGTM with some minor comments

**Comment by eyalfa** [comment]
@jrudolph thanks, I'll attend to these tonight.

**Comment by eyalfa** [comment]
@jrudolph , care to take one more look? thanks

**Review by jrudolph — APPROVED** [review]
LGTM (I committed the small typo fix I suggested).

**Comment by eyalfa** [comment]
@jrudolph , can u please merge this?

**Pull request merged**